practices/appeng/adrs/0002-distributed-tracing.md

```diff
@@ -0,0 +1,156 @@
+# 0002 - Use OpenTelemetry to instrument code for distributed tracing
+
+**Status:** In Progress
+
+**Date Superseded:** (if applicable)
+
+**Date Accepted:**
+
+**Reviewers:** @jim @reggieriser @rogeruiz \<tag yourself!>
+
+## Context
+
+### Objective
+
+_Give developers and operators an easier way to understand the behavior and
+structure of running systems by instrumenting code for distributed tracing._
+
+### Why Distributed Tracing?
+
+> Observability is defined as the ability of the internal states of a system to
+> be determined by its external outputs.
+
+There are several established techniques for gaining observability into complex
+software systems.
+
+- Logging
+  - helpful for known knowns
+  - e.g. "I want to know when a specific condition is reached"
+- Metrics
+  - helpful for assessing known unknowns
+  - e.g. "How many requests per second was the system handling last Tuesday?"
+- Distributed Tracing
+  - helpful for learning about unknown unknowns
+  - e.g. "What was the execution context for User X that caused their
+    interaction to timeout last Tuesday?"
+
+Some of the benefits of distributed tracing, as outlined in
+[this](https://petabridge.com/blog/why-use-distributed-tracing/) article are:
+
+- radically improves developer productivity and output
+- works across multiple applications, programming languages, and transports
+- improve time to market
+- facilitates excellent cross-team communication and cooperation
+
+Here are several example scenarios or questions that distributed tracing can
+help answer.
+
+- As a new engineer on the team, I want to understand how many separate systems
+  are involved when a certain user type logs in and the first page is rendered.
+- As an operations engineer, I want to know how many SQL queries are executed
+  for a given endpoint or interaction.
+- As a product manager, I want to know if a new feature is being used by a
+  certain cohort of users on a regular basis.
+- As an engineer, I want to prove that an optimization I wrote is effective
+  in a production environment.
+- As a load tester, after I have shown that a problem exists, I want to
+  understand how the system is interacting so I can debug and fix the issue.
+
+### ADR Goals and Anti-goals
+
+- Goal: Choose which set of libraries for generating
+  distributed tracing data to use at callsites (across programming
```

Comment:
What does callsites mean? Like does this mean the places within the code being instrumented? E.g. middleware, functions, etc.?

Reply:
I think it just means anywhere code is being called (this was language that came from the sourced MilMove PR).  Here's how Wikipedia [defines a call site](https://en.wikipedia.org/wiki/Call_site).  I'm open to alternative language here if you have a suggestion.

Reply:
I guess I would maybe suggest removing the call site bit since imo it doesn't add much. Like to me, something like

> Choose which set of libraries to use within a codebase (across programming languages) for generating distributed tracing data. 

gets the same point across, but maybe I'm missing something.